control-plane/subcommand/install-cni/cniconfig.go

```diff
@@ -0,0 +1,272 @@
+	consulMap, err := consulMapFromConfig(consulCfg)
+	if err != nil {
+		return fmt.Errorf("error converting consul config into map: %v", err)
+	}
+
+	// Append the consul-cni map to the already existing plugins.
+	cfgMap["plugins"] = append(plugins, consulMap)
+
+	// Marshal into a new json object
+	cfgJSON, err := json.MarshalIndent(cfgMap, "", "  ")
+	if err != nil {
+		return fmt.Errorf("error marshalling existing CNI config: %v", err)
+	}
+
+	// Libcni nuance/bug. If the newline is missing, the cni plugin will throw errors saying that it cannot get parse the config.
+	cfgJSON = append(cfgJSON, "\n"...)
+
+	// Write the file out.
+	err = os.WriteFile(cfgFile, cfgJSON, os.FileMode(0o644))
+	if err != nil {
+		return fmt.Errorf("error writing config file %s: %v", cfgFile, err)
+	}
+	return nil
+}
+
+// configFileToMap takes an unstructure JSON config file and converts it into a map.
+func configFileToMap(cfgFile string) (map[string]interface{}, error) {
+	if _, err := os.Stat(cfgFile); os.IsNotExist(err) {
+		return nil, fmt.Errorf("config file %s does not exist: %v", cfgFile, err)
+	}
+
+	// Read the main config file.
+	cfgBytes, err := os.ReadFile(cfgFile)
+	if err != nil {
+		return nil, fmt.Errorf("could not read file %s: %v", cfgFile, err)
+	}
+
+	// Convert the json config file into a map. The map that is created has 2 parts:
+	// [0] the cni header
+	// [1] the plugins
+	var cfgMap map[string]interface{}
+	err = json.Unmarshal(cfgBytes, &cfgMap)
+	if err != nil {
+		return nil, fmt.Errorf("error unmarshalling existing config file %s: %v", cfgFile, err)
+	}
+	return cfgMap, nil
+}
+
+// pluginsFromMap takes an unmarshalled config JSON map, return the plugin list asserted as a []interface{}.
+func pluginsFromMap(cfgMap map[string]interface{}) ([]interface{}, error) {
+	plugins, ok := cfgMap["plugins"].([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("error getting plugins from config map")
+	}
+	return plugins, nil
+}
+
+// consulMapFromConfig converts the consul CNI config into a map.
+func consulMapFromConfig(consulCfg *config.CNIConfig) (map[string]interface{}, error) {
+	var consulMap map[string]interface{}
+	err := mapstructure.Decode(consulCfg, &consulMap)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding consul config into a map: %v", err)
+	}
+	return consulMap, nil
+}
+
+// removeCNIConfig removes the consul-cni config from the CNI config file. Used as part of cleanup.
+func removeCNIConfig(cfgFile string) error {
+	// Read the config file and convert it to a map.
+	cfgMap, err := configFileToMap(cfgFile)
+	if err != nil {
+		return fmt.Errorf("could not convert config file to map: %v", err)
+	}
+
+	// Get the 'plugins' map embedded inside of the exisingMap.
+	plugins, err := pluginsFromMap(cfgMap)
+	if err != nil {
+		return err
+	}
+
+	// Find the 'consul-cni' plugin and remove it.
+	for i, p := range plugins {
+		plugin, ok := p.(map[string]interface{})
```

Comment:
could you add a comment here as to why we're not unmarshalling into `map[string]map[string]interface{}`

Reply:
Done.

```
// We do not unmarshall this into a map[string]map[string]interface{} because there is no structure to
// the plugins. They are unstructured json and can be in any format that the plugin provider wishes.
```